crates/symbolicator/src/cache.rs

```diff
@@ -201,23 +274,42 @@ impl Cache {
         // States a cache item can be in:
         // * negative/empty: An empty file. Represents a failed download. mtime is used to indicate
         //   when the failed download happened (when the file was created)
-        // * malformed: A file with the content `b"malformed"`. Represents a failed symcache
-        //   conversion. mtime indicates when we attempted to convert.
+        // * malformed: A file that starts with `b"malformed"`. Its contents should be one of the
+        //   following permutations:
+        //   - `b"malformed"` + b"some error description"
+        //   - `b"malformedbadobject"` + b"some error description"
+        //   - `b"malformeddownloaderror"` + b"some error description"
+        //   Like negative/empty, it represents a failed download and mtime represents when the
+        //   failure occurred. This covers all non-400 download errors.
         // * ok (don't really have a name): File has any other content, mtime is used to keep track
         //   of last use.
         let metadata = path.metadata()?;
 
         log::trace!("File length: {}", metadata.len());
 
-        let is_malformed = if MALFORMED_MARKER.len() as u64 <= metadata.len() {
+        // TODO: there's probably a cleaner way to write the entire block below
+        let (is_malformed, is_retryable) = if metadata.len() >= MALFORMED_MARKER.len() as u64 {
+            let max_to_read = MALFORMED_MARKER.len() + DOWNLOAD_ERROR_MARKER.len();
+            let amount_to_read = max_to_read.min(metadata.len() as usize);
+
             let mut file = File::open(path)?;
-            let mut buf = vec![0; MALFORMED_MARKER.len()];
+            let mut buf = vec![0; amount_to_read];
             file.read_exact(&mut buf)?;
 
             log::trace!("First {} bytes: {:?}", buf.len(), buf);
-            buf == MALFORMED_MARKER
+            if buf.starts_with(&[MALFORMED_MARKER, DOWNLOAD_ERROR_MARKER].concat())
```

Comment:
`.concat()` allocates, so we probably shouldn’t do it like this.

I can imagine something like this:

```
let mut is_retryable = false;
let is_malformed = buf.starts_with(...);
if is_malformed {
is_retryable = buf.starts_with(...) || buf.starts_with(...);
}
```

or similar. I think that is also easier to read than the `.concat()` variant.

Reply:
the logic here is pretty complex, so i've reworked this using a slightly different approach to track what type of cache file we have (look for `CacheStrategy` in the diff).